### PR TITLE
SPRINT: IODA-converters staging branch

### PR DESCRIPTION
**Pull request by rhoneyager**
## Description

This is the staging branch for the IODA-converters sprint of March 2022. This PR holds **IODA-converters** changes for individual converters.

When you create a PR in **Task 1** of the sprint, you should make a PR targeting this branch of the code.

For **Task 2**, your changes can target the ```develop``` branch directly.

## Errata

In this branch, the IODA validator is set to enforcing mode. This will help us determine which converters are still failing in CI.

**Note:** due to the CI system logic, your individual PRs will still be run in *permissive* mode. The CI tests will all pass. Only the ```feature/sprint-ioda-converters``` branch will log validation test failures.

### CODE REVIEW

File: CI/buildspec_clang.yml [hunk 1]
```diff
@@ -69,7 +69,7 @@ phases:
       - ./clone.sh $GIT_USER $GIT_PASS jcsda-internal/oops $CODEBUILD_GIT_BRANCH oops /jcsda/ioda-bundle develop
 
      # ioda
-      - ./clone.sh $GIT_USER $GIT_PASS jcsda-internal/ioda $CODEBUILD_GIT_BRANCH ioda /jcsda/ioda-bundle develop
+      - ./clone.sh $GIT_USER $GIT_PASS jcsda-internal/ioda $CODEBUILD_GIT_BRANCH ioda /jcsda/ioda-bundle feature/sprint-ioda-converters
```
Comment (rhoneyager) [inline CI/buildspec_clang.yml:72]: Note to self: before merging to develop, the buildspec_* files should be reverted.

**Comment by jeromebarre** [comment]
https://github.com/JCSDA-internal/ioda-converters/pull/850 hasn't been merged I think. Looking at the file change nothing on viirs aod.

**Comment by srherbener** [comment]
Could someone please look at these conflicts. It's not clear to me how they should be resolved. Thanks!

File: src/conventional/amv_ssec_ascii2ioda.py [hunk 2]
```diff
@@ -179,18 +186,25 @@ def read_file(file_name, data):
...
-                year = int(row['day'][0:4])
-                month = int(row['day'][5:6])
-                day = int(row['day'][7:8])
-                hour = int(row['hms'][0:1])
-                minute = int(row['hms'][2:3])
-                second = 0
-                dtg = datetime(year, month, day, hour, minute, second)
+                dtg = datetime.strptime(f"{row['day']} {row['hms']}", '%Y%m%d %H%M')
                 time_offset = np.int64(round((dtg - epoch).total_seconds()))
                 data['dateTime'] = np.append(data['dateTime'], time_offset)
```
Comment (haydenlj) [inline src/conventional/amv_ssec_ascii2ioda.py:191]: the merge conflict appears to be with the performance upgrade to this converter that was merged in (https://github.com/JCSDA-internal/ioda-converters/pull/934) I don't have permission to push any changes but lines 191-246 should be changed to:

 ```
                local_data['dateTime'] = np.append(local_data['dateTime'], time_offset)
                local_data['longitude'] = np.append(local_data['longitude'], float(row['lon']))
                local_data['latitude'] = np.append(local_data['latitude'], float(row['lat']))

                pres = float(row['pre'])*100.
                local_data['air_pressure'] = np.append(local_data['air_pressure'], pres)
                wdir = float(row['dir'])*1.0
                wspd = float(row['spd'])*1.0
                if (wdir >= 0 and wdir <= 360 and wspd >= 0 and wspd < 300):
                    uwnd, vwnd = meteo_utils.meteo_utils().dir_speed_2_uv(wdir, wspd)
                else:
                    uwnd = float_missing_value
                    vwnd = float_missing_value

                local_data['eastward_wind'] = np.append(local_data['eastward_wind'], uwnd)
                local_data['northward_wind'] = np.append(local_data['northward_wind'], vwnd)

                if row['type'] in known_freq.keys():
                    freq = known_freq[row['type']]
                else:
                    freq = float_missing_value
                    unk_freq.append(row['type'])
                local_data['sensorCentralFrequency'] = np.append(local_data['sensorCentralFrequency'], freq)

                if row['sat'] in known_sat.keys():
                    satid = known_sat[row['sat']]
                else:
                    satid = int_missing_value
                    unk_sat.append(row['sat'])

                local_data['satelliteIdentifier'] = np.append(local_data['satelliteIdentifier'], satid)
                local_data['sensorZenithAngle'] = np.append(local_data['sensorZenithAngle'], float(row['rff']))
                local_data['windTrackingCorrelation'] = np.append(local_data['windTrackingCorrelation'], float(row['qi']))
                local_data['windHeightAssignMethod'] = np.append(local_data['windHeightAssignMethod'], int(row['int']))

            except KeyError as e:
                if not keyerr:
                    logging.warning(file_name + ' is missing variable ' + e.args[0])
                keyerr = True

                if (e.args[0] == 'dir') or (e.args[0] == 'spd'):
                    local_data['eastward_wind'] = np.append(local_data['eastward_wind'], float_missing_value)
                    local_data['northward_wind'] = np.append(local_data['northward_wind'], float_missing_value)
                else:
                    local_data[known_var[e.args[0]][0]] = np.append(local_data[known_var[e.args[0]][0]], known_var[e.args[0]][1])

        for f in unk_freq:
            logging.warning(file_name + ' contains unknown frequency ' + f)

        for s in unk_sat:
            logging.warning(file_name + ' contains unknown satellite ID ' + s)

        if len(row.keys()) > len(known_var.keys()):
            for k in row.keys():
                if k not in known_var.keys():
                    logging.warning(file_name + ' contains unknown variable ' + k)

    for key in varDict.keys():
        data[key] = np.append(data[key], local_data[key])
    for key in meta_keys:
        data[key] = np.append(data[key], local_data[key])

```

**Comment by YoulongXia-NOAA** [comment]
@rhoneyager @haydenlj , I checked ioda-converter code sprint work table, this converter was done by Greg Thomson. Please contact him and see if he can fix.

**Comment by srherbener** [comment]
> @rhoneyager @haydenlj , I checked ioda-converter code sprint work table, this converter was done by Greg Thomson. Please contact him and see if he can fix.

@haydenlj thank you for the instructions. I'll resolve those conflicts since I have permission to change the `feature/sprint-ioda-converters` branch. Greg might have the same issue with not being able to modify the `feature/sprint-ioda-converters` branch.

**Comment by srherbener** [comment]
@haydenlj your instructions worked but there are still some conflicts with old and new variable names. However, it's clear what should be done to resolve these so I can take care of that.

@YoulongXia-NOAA are you the person who could give instructions on how to resolve the conflicts with the `src/land/smap9km_ssm2ioda.py` script? Thanks!

**Comment by YoulongXia-NOAA** [comment]
> @haydenlj your instructions worked but there are still some conflicts with old and new variable names. However, it's clear what should be done to resolve these so I can take care of that.
> 
> @YoulongXia-NOAA are you the person who could give instructions on how to resolve the conflicts with the `src/land/smap9km_ssm2ioda.py` script? Thanks!

@srherbener, I double check the script and it looks good to me. Thank you for fix the conflict. Please make a test run and update the reference.

**Comment by srherbener** [comment]
> > @haydenlj your instructions worked but there are still some conflicts with old and new variable names. However, it's clear what should be done to resolve these so I can take care of that.
> > @YoulongXia-NOAA are you the person who could give instructions on how to resolve the conflicts with the `src/land/smap9km_ssm2ioda.py` script? Thanks!
> 
> @srherbener, I double check the script and it looks good to me. Thank you for fix the conflict. Please make a test run and update the reference.

Thanks for checking the script. I'll let the current testing finish and then update the reference if need be.

**Comment by srherbener** [comment]
@YoulongXia-NOAA, updating the test reference got the smap9k test to pass, but now the validator fails on the smap9k reference file:
```
Reading YAML from /jcsda/ioda-bundle/ioda/share/ioda/yaml/validation/ObsSpace.yaml
Processing data file: /jcsda/ioda-bundle/iodaconv/test/testoutput/smap9km_ssm.nc
 Verifying that all required groups exist
 Verifying group MetaData
 Verifying group ObsValue
 Verifying group ObsError
 Verifying group PreQC
 Verifying group /
 Verifying dimension names
 Verifying variable information
  Variable MetaData/dateTime
  Variable MetaData/latitude
  Variable MetaData/longitude
  Variable MetaData/vegetationOpacity
   Warning: Attribute 'units' is present but is not listed as a required or optional attribute in the spec.
   ERROR: Variable 'MetaData/vegetationOpacity' needs units, but the 'units' attribute does not exist in the YAML.
  Variable MetaData/earthSurfaceType
  Variable MetaData/easeRowIndex
   Warning: Variable 'MetaData/easeRowIndex' is not listed in the YAML conventions file.
  Variable MetaData/easeColumnIndex
   Warning: Variable 'MetaData/easeColumnIndex' is not listed in the YAML conventions file.
  Variable ObsValue/soilMoistureVolumetric
  Variable ObsError/soilMoistureVolumetric
  Variable PreQC/soilMoistureVolumetric
Final results:
  # errors:      1
  # warnings:    3

CTEST_FULL_OUTPUT
```

Can you help get this fixed? Thanks!

**Comment by YoulongXia-NOAA** [comment]
> vegetationOpacity

@srherbener, I checked both  ObsSpace.yaml and convention table, vegetationOpacity has no units.  you need to remove 
self.varAttrs[('vegetationOpacity', 'MetaData')]['units'] = 'unitless' from smap9km_ssm2ioda.py. I forgot remove it.

For MetaData/easeRowIndex and MetaData/easeColumnIndex, I need to add them to ObsSpace.yaml . I will work them soon.

**Comment by YoulongXia-NOAA** [comment]
@srherbener, by the way, if I update ObsSpace.yaml , do I need to use  feature/sprint-ioda-converters or develop as the reference branch? After I get your response, I will work it soon.

**Comment by srherbener** [comment]
> > vegetationOpacity
> 
> @srherbener, I checked both ObsSpace.yaml and convention table, vegetationOpacity has no units. you need to remove self.varAttrs[('vegetationOpacity', 'MetaData')]['units'] = 'unitless' from smap9km_ssm2ioda.py. I forgot remove it.
> 
> For MetaData/easeRowIndex and MetaData/easeColumnIndex, I need to add them to ObsSpace.yaml . I will work them soon.

@YoulongXia-NOAA I have pushed to github a new feature brach, `feature/sprint-smap9k-cleanup`, with the fix for the vegetationOpacity variable (ie, I removed the units attribute from this variable). I've updated the test reference file accordingly and we are now down to the two warnings from the validator (which are the two new variables, easeRowIndex and easeColumnIndex). 

The new `feature/sprint-smap9k-cleanup` branch is based on the `feature/sprint-ioda-converters` branch. Please make your additions to ObsSpace.yaml on the new `feature/sprint-smap9k-cleanup` branch and then submit a PR to the `feature/sprint-ioda-converters` branch. Thanks!

**Comment by YoulongXia-NOAA** [comment]
@srherbener, okay, got it. Than you.

**Comment by srherbener** [comment]
@FanHan-NOAA I think that the conflict needs to be resolved by making the same changes in the test reference file `test/testoutput/sfc_tv_obs_2018041500.nc4` that were done in #953. Thanks!

**Review by PatNichols — APPROVED** [review]

**Review by BenjaminRuston — APPROVED** [review]
well at least it wasn't a year Ryan ;0p

**Review by gthompsnJCSDA — APPROVED** [review]
Hurray, we are here finally!

**Pull request merged**